### PR TITLE
Refactor filtering of selected signals and default VSS tree loading to --no-expand

### PR DESCRIPTION
**Pull request by Kostadin-Ivanov**
Updates done:

Refactored vss2samm exporter to load VSS Tree only in **--no-expand** mode - this was required so to avoid redundancy in created aspect model nodes, when the tree was expanded, by default. In other words, when the VSS instances were instantiated by default. Thus, the need to checks like: **is_node_expanded** was no longer needed so related code for these checks was also cleaned.

### CODE REVIEW

**Comment by Kostadin-Ivanov** [comment]
Hello @sschleemilch ,

here is refactored version of filtering of the VSS Tree and setting the vss2samm to load only --no-expand tree.

**Comment by sschleemilch** [comment]
What is the initial reason that you want to allow incomplete paths to be specified?
The instances concept in the end is just to need to write less lines in the *.vspec files but the truth is the expanded tree and in my opinion also what users should see. Instead of specifying non expanded paths it would maybe make more sense to let users define signals using wildcards such as  `Vehicle.Cabin.Door.*.*.Window` which is probably way easier to process then instead of trying to re-implement the instance logic (which is complicated enough doing it once)

So I am not sure it is a good idea to provide users the interface to the non expanded tree.
I think there is a lot of complexity implemented here which in the end re-implements the instance expansion logic.

**Comment by Kostadin-Ivanov** [comment]
> What is the initial reason that you want to allow incomplete paths to be specified? The instances concept in the end is just to need to write less lines in the *.vspec files but the truth is the expanded tree and in my opinion also what users should see. Instead of specifying non expanded paths it would maybe make more sense to let users define signals using wildcards such as `Vehicle.Cabin.Door.*.*.Window` which is probably way easier to process then instead of trying to re-implement the instance logic (which is complicated enough doing it once)
> 
> So I am not sure it is a good idea to provide users the interface to the non expanded tree. I think there is a lot of complexity implemented here which in the end re-implements the instance expansion logic.

The expanded tree is causing redundancies in the generated aspect models. 
I tried to explain this a couple of times with the Door example, having 4 DoorEntity trees for each instance: Row1/2 Driver and Passenger side.

Currently the logic for the instantiation of the door instances in the vss2samm exporter allows to have a DoorInstanceEntity with Row1/2 Driver / Passenger side branches, where each of these is pointing to a DoorEntity which represents the proper Door tree (branch) with its details, signals and sub-branches.

I am sorry that, not sure why, I cannot attach screenshots here so to visualize what I mean.

Just for another example, the aspect model for "not expanded" Cabin tree has 483 elements. 
It if was expanded, then each branch which had any instances, would have doubled for 2 instances (left and right) and quadrupled for instances with 1,2 rows and left/right side, which results in 1508 elements. 

This is more than triple the "not expanded" version of the Cabin aspect model.

For sure, we can try to refactor the generation of these so they is no such redundancy when VSS tree is expanded, but for the moment we don't have the time and resources for this.

File: src/vss_tools/vspec/tree.py
Comment (Kostadin-Ivanov) [reply src/vss_tools/vspec/tree.py:274]: When this was in INFO mode, my console got flooded with below message:

`INFO     Nodes deleted, given=1, overall=0`

which on my opinion still does not say much i.e., which node got deleted from where.

By the way, I am calling the deletion, only once in the **filter_vss_tree** function. 
Actually, the function is calling itself recursively for each child node, so the **delete_nodes** gets called on each execution of **filter_vss_tree**.
Anyway, I took it outside of the **filter_vss_tree** function, and called the **delete_nodes** in the vss2samm.py, rjust after the **filter_vss_tree** function completed with marking of nodes for deletion, and still got the same output in the console. Exactly the same as above INFO message "flooded" my console again.

Anyway, as you want to see it, then I will rollback this change.

File: src/vss_tools/vspec/vssexporters/vss2samm/vss2samm.py
Comment (Kostadin-Ivanov) [reply src/vss_tools/vspec/vssexporters/vss2samm/vss2samm.py:249]: I reworked this and cleared the ignore.

File: src/vss_tools/vspec/tree.py
Comment (sschleemilch) [reply src/vss_tools/vspec/tree.py:274]: if overall = 0 then something is odd since you should only delete nodes that are reachable from the node you are calling the delete.

I agree that when calling this more than  once that the output is not good or helpful.
Right now it was only called once: for the vss tree that has been loaded

**Comment by sschleemilch** [comment]
> > What is the initial reason that you want to allow incomplete paths to be specified? The instances concept in the end is just to need to write less lines in the *.vspec files but the truth is the expanded tree and in my opinion also what users should see. Instead of specifying non expanded paths it would maybe make more sense to let users define signals using wildcards such as `Vehicle.Cabin.Door.*.*.Window` which is probably way easier to process then instead of trying to re-implement the instance logic (which is complicated enough doing it once)
> > So I am not sure it is a good idea to provide users the interface to the non expanded tree. I think there is a lot of complexity implemented here which in the end re-implements the instance expansion logic.
> 
> The expanded tree is causing redundancies in the generated aspect models. I tried to explain this a couple of times with the Door example, having 4 DoorEntity trees for each instance: Row1/2 Driver and Passenger side.
> 
> Currently the logic for the instantiation of the door instances in the vss2samm exporter allows to have a DoorInstanceEntity with Row1/2 Driver / Passenger side branches, where each of these is pointing to a DoorEntity which represents the proper Door tree (branch) with its details, signals and sub-branches.
> 
> I am sorry that, not sure why, I cannot attach screenshots here so to visualize what I mean.
> 
> Just to provide some real figures :) The aspect model for "not expanded" **Cabin tree** has 483 elements. It if was expanded, then each branch which had any instances, would have doubled for 2 instances (left and right) and quadrupled for instances with 1,2 rows and left/right side, which results in 1508 elements.
> 
> This is more than triple the "not expanded" version of the Cabin aspect model.
> 
> For sure, we can try to refactor the generation of these so they is no such redundancy when VSS tree is expanded, but for the moment we don't have the time and resources for this.

Okay, I think I understood a bit better what it is about and yeah, probably acting on the non expanded tree is exactly what you need for the generation. So all  good

File: src/vss_tools/vspec/vssexporters/vss2samm/vss2samm.py
Comment (Kostadin-Ivanov) [reply src/vss_tools/vspec/vssexporters/vss2samm/vss2samm.py:247]: Done

File: src/vss_tools/vspec/vssexporters/vss2samm/helpers/vss_helper.py
Comment (Kostadin-Ivanov) [reply src/vss_tools/vspec/vssexporters/vss2samm/helpers/vss_helper.py:178]: The idea was to check if current vss_path is part of the sdelected_path and vice versa, instead of if one starts with the other, but you are right.

I reworked this check as advised i.e. to check if one path starts with the other or vice versa.

Done.

File: src/vss_tools/vspec/vssexporters/vss2samm/helpers/vss_helper.py
Comment (Kostadin-Ivanov) [reply src/vss_tools/vspec/vssexporters/vss2samm/helpers/vss_helper.py:196]: Done

File: src/vss_tools/vspec/vssexporters/vss2samm/helpers/vss_helper.py
Comment (Kostadin-Ivanov) [reply src/vss_tools/vspec/vssexporters/vss2samm/helpers/vss_helper.py:214]: I tried this, as mentioned in my comment below and:

>I took it outside of the filter_vss_tree function, and called the delete_nodes in the vss2samm.py, rjust after the filter_vss_tree function completed with marking of nodes for deletion, and still got the same output in the console. Exactly the same as above INFO message "flooded" my console again.

Anyway, I'd say that your suggestion to call the delete after we have "filtered" the whole tree i.e. outside of the filter_vss_tree is better, so refactored this code as advised.

Done.

File: src/vss_tools/vspec/tree.py
Comment (Kostadin-Ivanov) [reply src/vss_tools/vspec/tree.py]: Thank you, I got it :)

After I reworked the deletion and move the call to delete, outside the filter function, I got just 1 message stating just deleted nodes which was more than enough :)

**Review by sschleemilch — APPROVED** [review]

**Comment by Kostadin-Ivanov** [comment]
Thank you @sschleemilch .

I am not sure if I need another approval, or this can be merged now. Also, I think that I don't have permissions to merge it right?

Maybe @erikbosch or you, Sebastian, can do it?

**Comment by sschleemilch** [comment]
I don't have merge persmissions 😿

**Comment by Kostadin-Ivanov** [comment]
So it is Erik then :)

**Comment by erikbosch** [comment]
Yes - It is typically me, Adnan or Sebastian Schildt that press the merge button. (I guess Paul Boyes and some other also have "general" organization level rights on COVESA and can also theoretically do it. )

As this only concerns SAMM, which is a new tool and a peer reviewed has happened I see no reason to wait for next meeting, so I will just do a quick sanity check,

**Review by erikbosch — APPROVED** [review]
LGTM

**Comment by Kostadin-Ivanov** [comment]
Thank you very much Erik.

Now we just need to wait for the release :)

**Pull request merged**
